lib/spack/spack/buildcache_prune.py

```diff
@@ -248,24 +268,145 @@ def _prune_orphans(
     return pruned_object_count
 
 
-def prune(mirror: Mirror, dry_run: bool) -> None:
+def prune_direct(
+    mirror: Mirror, keeplist_file: pathlib.Path, pruning_started_at: float, dry_run: bool
+) -> None:
+    """
+    Execute direct pruning for a given mirror using a keeplist file.
+
+    This function reads a file containing spec hashes to keep, then deletes
+    all other spec manifests from the buildcache.
+    Note that this function does *not* prune the blobs associated with the manifests;
+    to do that, `prune_orphan` must be invoked to clean up the now-orphaned blobs.
+
+    Args:
+        mirror: Mirror to prune
+        keeplist_file: Path to file containing newline-delimited hashes to keep
+        pruning_started_at: Timestamp of when the pruning started
+        dry_run: Whether to perform a dry run without actually deleting
+    """
+    tty.info("=== Direct Pruning Phase ===")
+    tty.debug(f"Direct pruning mirror: {mirror.fetch_url}" + (" (dry run)" if dry_run else ""))
+
+    keep_hashes: Set[str] = set()
+    for line in keeplist_file.read_text().splitlines():
+        keep_hash = line.strip()
```

Comment:
It may also make sense to support the case where the hash starts with `/` since that is often how Spack formats hashes.

```suggestion
        keep_hash = line.strip()
        if keep_hash[0] == "/":
            keep_hash = keep_hash[1:]
```

Reply:
https://github.com/spack/spack/pull/51153/commits/e98eec8599f0a662494eccee236d0373330aec04